### ISSUE TITLE
Unpacking ends prior to end of file.

### ISSUE THREAD

**Issue by ksmith0** [pixie16/paass#201]
I believe this issue may be related to ScanLib, but may have already been resolved. The following is encountered when scanning a large background run (>2 GB) with the histPlotter branch:
```
$ hist -i background_001.ldf --batch
hist: Initializing derived class.

 ScanInterface v1.2.29
 ==  ==  ==  ==  == 

hist: Using filename background_001.ldf.
 'DIR ' buffer-
  Run number: 1
  Number buffers: 68168
 'HEAD' buffer-
  Facility: HHIRF   
  Format: L003    
  Type: LIST DATA       
  Date: 01/30/17 17:10  
  Title: HAGRiD det. 201 background (CAEN crate @ -1100V)                                
  Run number: 1

hist: File loaded.
[READ] 65562 words (0%), GOOD = 9, LOST = 0BuildRawEvent: First event time is 6.86414e+11 clock ticks.
[READ] 65562 words (1%), GOOD = 819, LOST = 0
[READ] 65562 words (2%), GOOD = 1612, LOST = 0
[READ] 65562 words (3%), GOOD = 2413, LOST = 2
[READ] 65562 words (4%), GOOD = 3206, LOST = 2
[READ] 65562 words (5%), GOOD = 3836, LOST = 2
[READ] 65508 words (6%), GOOD = 4638, LOST = 2
[READ] 65616 words (7%), GOOD = 5431, LOST = 2
[READ] 65562 words (8%), GOOD = 6224, LOST = 2
[READ] 65562 words (9%), GOOD = 6980, LOST = 2

hist: Scan complete.
[STOP] Acquisition stopped.
Running ScanInterface v1.2.29 (Aug. 11th, 2016)
hist: Retrieved 775 spills!

hist: Cleaning up...
hist: Read 6989 spill chunks.
hist: Lost at least 2 spill chunks.
```

The initial investigation should involve rebasing dev onto histPlotter (is that the correct order) to check if the issue has not already been resolved upstream.

**Comment by spaulaus** [comment]
I've seen this when taking data during Shloka's visit. I believe it was resolved in Pull Request #166

**Comment by ksmith0** [comment]
Thanks for the hint. Hopefully it can be looked at today.

**Comment by cthornsb** [comment]
I propagated the changes in pull 166 (hribf_fixes) to my local version to work on this problem. Using debug mode, I identified which buffer in the ldf file was causing problems. The error is shown below.

![image](https://cloud.githubusercontent.com/assets/7407864/22477242/59a358e0-e7b3-11e6-84c5-dec46addb5ea.png)

The ldf writer correctly placed the spill footer in the next buffer because the top one only had 4 words left and the spill footer requires at least 5 free words. That worked correctly. However, it didn't add enough 0xFFFFFFFF to the end of the top buffer to close it. So the top buffer ended up being short by 4 words.

I have seen this error manifest more in files taken when the polling threshold is really low (~1%). I believe the reason for this is that we are writing smaller spills to file and are more likely to encounter this problem.

**Comment by cthornsb** [comment]
I believe this is the culprit.

![image](https://cloud.githubusercontent.com/assets/7407864/22477521/6ae487ea-e7b4-11e6-9c0e-6c59e6e7395f.png)

I think adding a "Close(file_)" before the open_(file_) should fix the problem, but I haven't had time to test it yet.

***NOTE*** DATA_buffer::Close is the method responsible for filling the current buffer to 8194 words.

**Comment by ksmith0** [comment]
Why would this affect the file at 9%? Why is there an open here at all?

**Comment by ksmith0** [comment]
This may be related to #107.

**Comment by cthornsb** [comment]
If the spill footer will not fit in the remaining space in the buffer, it opens a new buffer. It's an error that it doesn't close the current one first. So the buffer reader gets out of position because it expects 8194 words but there are less than 8194 words in the buffer. It is evidently a very rare occurrence, but the problem seems to get worse when we're writing more spill footers (i.e. when the polling threshold is lowered).

**Comment by cthornsb** [comment]
In addition to fixing this problem, we should probably change the DATA buffer reader to scan for the next DATA word if it encounters a non-DATA buffer.

**Comment by cthornsb** [comment]
With a polling threshold of 0.1% (131 words), I just took four 90 second data files using a HAGRiD detector w/ no source (~200 Hz) using the up-to-date dev branch. Every one of the files exhibited this problem. Apparently it is very easy to reproduce. I'm going to make the change and see if it fixes it.

**Comment by cthornsb** [comment]
All of the test files I took with the Close() fix no longer exhibit the problem.

![image](https://cloud.githubusercontent.com/assets/7407864/22517800/56ee6232-e878-11e6-8ca7-e9cfa30f80ef.png)

The screenshot above shows that a buffer was closed with three delimiters (one more than usual) and the five word spill footer was placed in the next buffer.

I have pushed this fix to a new branch "hribf_fixes".

**Comment by spaulaus** [comment]
Is there a way to fix files that have been affected by this issue?

**Comment by ksmith0** [comment]
Resolved in commit b5d2036ca0ac4370fbdb67ab5f8fe63e20a0d204

**Comment by cthornsb** [comment]
Brute force method probably. Make a copy of the file word by word and add extra delimiters wherever needed. Such a program does not exist yet.
